rclpy/src/rclpy/_rclpy.c

```diff
@@ -2983,6 +2984,250 @@ rclpy_get_rmw_qos_profile(PyObject * Py_UNUSED(self), PyObject * args)
+ * \return NULL on failure:
+ *         PyLong integer in nanoseconds on success
+ */
+static PyObject *
+rclpy_time_point_get_nanoseconds(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * pytime_point;
+  if (!PyArg_ParseTuple(args, "O", &pytime_point)) {
+    return NULL;
+  }
+
+  rcl_time_point_t * time_point = (rcl_time_point_t *)PyCapsule_GetPointer(
+    pytime_point, "rcl_time_point_t");
+  if (!time_point) {
+    return NULL;
+  }
+
+  return PyLong_FromUnsignedLongLong(time_point->nanoseconds);
+}
+
+/// Destructor for a duration
+void
+_rclpy_destroy_duration(PyObject * pycapsule)
+{
+  PyMem_Free(PyCapsule_GetPointer(pycapsule, "rcl_duration_t"));
+}
+
+/// Create a duration
+/**
+ * On failure, an exception is raised and NULL is returned if:
+ *
+ * Raises RuntimeError on initialization failure
+ * Raises TypeError if argument of invalid type
+ * Raises OverflowError if nanoseconds argument cannot be converted to uint64_t
+ *
+ * \param[in] nanoseconds unsigned PyLong object storing the nanoseconds value
+ *   of the duration in a 64-bit unsigned integer
+ * \return Capsule of the pointer to the created rcl_duration_t * structure, or
+ * \return NULL on failure
+ */
+static PyObject *
+rclpy_create_duration(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * pylong_nanoseconds;
+
+  if (!PyArg_ParseTuple(args, "O", &pylong_nanoseconds)) {
+    return NULL;
+  }
+
+  unsigned PY_LONG_LONG nanoseconds = PyLong_AsUnsignedLongLong(pylong_nanoseconds);
+  if (PyErr_Occurred()) {
+    return NULL;
+  }
+
+  rcl_duration_t * duration = (rcl_duration_t *) PyMem_Malloc(sizeof(rcl_duration_t));
+  if (NULL == duration) {
+    PyErr_Format(PyExc_RuntimeError, "Failed to allocate memory for duration.");
+    return NULL;
+  }
+
+  duration->nanoseconds = nanoseconds;
+
+  return PyCapsule_New(duration, "rcl_duration_t", _rclpy_destroy_duration);
+}
+
+/// Returns the nanoseconds value of the duration
+/**
+ * On failure, an exception is raised and NULL is returned if:
+ *
+ * Raises ValueError if pyduration is not a duration capsule
+ *
+ * \param[in] pyduration Capsule pointing to the duration
+ * \return NULL on failure:
+ *         PyLong integer in nanoseconds on success
+ */
+static PyObject *
+rclpy_duration_get_nanoseconds(PyObject * Py_UNUSED(self), PyObject * args)
+{
+  PyObject * pyduration;
+  if (!PyArg_ParseTuple(args, "O", &pyduration)) {
+    return NULL;
+  }
+
+  rcl_duration_t * duration = (rcl_duration_t *)PyCapsule_GetPointer(
+    pyduration, "rcl_duration_t");
+  if (!duration) {
+    return NULL;
+  }
+
+  return PyLong_FromUnsignedLongLong(duration->nanoseconds);
+}
+
+/// Destructor for a clock
+void
+_rclpy_destroy_clock(PyObject * pycapsule)
+{
+  rcl_clock_t * clock = (rcl_clock_t *)PyCapsule_GetPointer(pycapsule, "rcl_clock_t");
+  rcl_ret_t ret_clock = rcl_clock_fini(clock);
+  PyMem_Free(clock);
+  if (ret_clock != RCL_RET_OK) {
+    PyErr_Format(PyExc_RuntimeError,
```

Comment:
@sloretz I had overlooked that I wasn't calling fini on the clock (added in 0cbbc52b3ceeb4420ae2148c97e64f80b6f05243).

If something goes wrong in the fini when it gets garbage collected it will cause this sort of output:

```
RuntimeError: Failed to fini 'rcl_clock_t': error not set

During handling of the above exception, another exception occurred:

self = <contextlib._GeneratorContextManager object at 0x7f1210674cf8>, type = None, value = None, traceback = None

    def __exit__(self, type, value, traceback):
        if type is None:
            try:
>               next(self.gen)
E               SystemError: PyEval_EvalFrameEx returned a result with an error set

/usr/lib/python3.5/contextlib.py:66: SystemError
----------- generated xml file: /home/dhood/ros2_ws/build/rclpy/test_results/rclpy/rclpytests.xunit.xml -----------
============================================ 2 failed in 0.06 seconds =============================================
Exception ignored in: 'garbage collection'
RuntimeError: Failed to fini 'rcl_clock_t': error not set
Fatal Python error: unexpected exception during garbage collection

Current thread 0x00007f1216872700 (most recent call first):
Aborted (core dumped)
```

I think if there was an issue then this will certainly let the user know about it. Since you were the one that mentioned adding the destructor for garbage collection, I wonder if you've read about any other approaches to signalling issues in the destructor, or if this is the way to go?

Reply:
Wow, the user can't miss that. Does [`PyErr_WarnEx()`](https://docs.python.org/3/c-api/exceptions.html#c.PyErr_WarnEx) cause a fatal error as well?

I haven't read about other approaches. The only example I found in the cpython repo is [this capsule destructor in ctypes](https://github.com/python/cpython/blob/e2f92de6a90ae11a6d8e01bd17fd88b005516835/Modules/_ctypes/callproc.c#L87-L93) which is seems to be silent about errors.

Reply:
Using `PyErr_WarnFormat`, you end up getting something like:

```
----------- generated xml file: /home/dhood/ros2_ws/build/rclpy/test_results/rclpy/rclpytests.xunit.xml -----------
================================================ warnings summary =================================================
src/ros2/rclpy/rclpy/test/test_clock.py::TestClock::test_clock_construction
  /home/dhood/ros2_ws/src/ros2/rclpy/rclpy/test/test_clock.py:33: RuntimeWarning: Failed to fini 'rcl_clock_t': error not set
    clock = Clock(clock_type=ClockType.STEADY_TIME)
  /home/dhood/ros2_ws/src/ros2/rclpy/rclpy/test/test_clock.py:35: RuntimeWarning: Failed to fini 'rcl_clock_t': error not set
    clock = Clock(clock_type=ClockType.SYSTEM_TIME)
  /usr/lib/python3.5/unittest/case.py:600: RuntimeWarning: Failed to fini 'rcl_clock_t': error not set
    testMethod()

src/ros2/rclpy/rclpy/test/test_clock.py::TestClock::test_clock_now
  /home/dhood/ros2_ws/src/ros2/rclpy/rclpy/test/test_clock.py:53: RuntimeWarning: Failed to fini 'rcl_clock_t': error not set
    clock = Clock(clock_type=ClockType.STEADY_TIME)
  /usr/lib/python3.5/unittest/case.py:600: RuntimeWarning: Failed to fini 'rcl_clock_t': error not set
    testMethod()

-- Docs: http://doc.pytest.org/en/latest/warnings.html
====================================== 2 passed, 5 warnings in 0.03 seconds =======================================
```

But I think now it's misleading because it doesn't make mention of the garbage collector so the line numbers are confusing. We do raise exceptions if there are [fini errors usually](https://github.com/ros2/rclpy/blob/98ad24ae08f553c46c683e98350333cd0bda42cd/rclpy/src/rclpy/_rclpy.c#L2022), so if there aren't any alternatives in between, now I think that the exception is the right thing even if it's perhaps a bit extreme.